packages/core/src/common/disposable.spec.ts

```diff
@@ -26,4 +29,72 @@ describe('Disposables', () => {
         expect(collectionA.disposed, 'A should be disposed after being disposed.').to.be.true;
         expect(collectionB.disposed, 'B should not be disposed because A was disposed.').to.be.false;
     });
+
+    it('Collection is auto-pruned when an element is disposed', () => {
+        const onDispose = spy(() => { });
+        const elementDispose = () => { };
+
+        const collection = new DisposableCollection();
+        collection.onDispose(onDispose);
+
+        // DisposableCollection doesn't provide direct access to its array,
+        // so we have to defeat TypeScript to properly test pruning.
+        function collectionSize(): number {
```

Comment:
I'm not the biggest fan of this construct and I believe we could replace this by simply accessing the `disposables` and utilizing the [lenghtOf](https://www.chaijs.com/api/bdd/#method_lengthof) method from chai.

```
expect(collection['disposables']).to.have.lengthOf(1);
expect(collection['disposables']).to.be.empty;
```

Reply:
Oh gosh. I'm still such a TypeScript newbie in many ways. I had no idea that accessing a property with bracket notation circumvents the type's accessibility attributes!

Reply:
Haha, yes, we can profiting from the freedom of JavaScript and indexed access!

Reply:
Well, it's funny because at first I thought backet access would give you an `any` and give you full freedom from there on out. But not so! 
```
class SomeThing {
    foo() : void {}
    private bar() : void {}
}
class SomeClass {
    private obj = new SomeThing();
}
const sc = new SomeClass();
sc.obj.foo();      // error; obj is private
sc['obj'].foo();   // ok; bracket notation to the rescue
sc['jjj'].foo();   // error property jjj does not existin in SomeClass
sc['obj'].bar();   // error; compiler doesn't treat sc['obj'] as any.
                   // bracket notation circumvents accessibility
                   // traits but that's about it, I think
````

Reply:
You are right, index access does not imply an implicit `any`. This is definitely an advantage as you keep your type-safety for compiler checks (e.g., someone renames the field would result in a compile error instead of a runtime error). So in your example you would still be able to do `sc['obj']['bar']();` to call the private method of a private object and get all the typing information.